File: src/main/java/net/zero918nobita/Xemime/entity/Bool.java
package net.zero918nobita.Xemime.entity;

import net.zero918nobita.Xemime.NodeType;
import net.zero918nobita.Xemime.ast.FatalException;
import net.zero918nobita.Xemime.ast.Node;

public class Bool extends Node {
    private boolean bool;

    public static Bool T = new Bool(true);
    public static Bool Nil = new Bool(false);

    public Bool(int location, boolean bool) {
        super(location);
        this.bool = bool;
    }

    @Override
    public NodeType recognize() {
        return NodeType.BOOL;
    }

    public Bool(boolean bool) {
        this(0, bool);
    }

    @Override
    public String toString() {
        return (bool) ? "T" : "NIL";
    }

    @Override
    public boolean equals(Object obj) {
        return (obj instanceof Bool) && (((Bool)obj).isTrue() == this.bool);
    }

    public boolean isTrue() {
        return bool;
    }

    @Override
    public Bool and(int location, Node rhs) throws FatalException {
        if (rhs instanceof Bool) return (bool && (((Bool)rhs).isTrue())) ? Bool.T : Bool.Nil;
        else throw new FatalException(location,  117);
    }

    @Override
    public Bool or(int location, Node rhs) throws FatalException {
        if (rhs instanceof Bool) return (bool || (((Bool) rhs).isTrue())) ? Bool.T : Bool.Nil;
        else throw new FatalException(location, 118);
    }

    @Override
    public Bool xor(int location, Node rhs) throws FatalException {
        if (rhs instanceof Bool) return (bool ^ (((Bool) rhs).isTrue())) ? Bool.T : Bool.Nil;
        else throw new FatalException(location, 119);
    }
}

File: src/main/kotlin/net/zero918nobita/xemime/entity/Bool.kt
package net.zero918nobita.xemime.entity

import net.zero918nobita.xemime.NodeType
import net.zero918nobita.xemime.ast.FatalException
import net.zero918nobita.xemime.ast.Node
import kotlin.Int

class Bool(location: Int, private val bool: Boolean) : Node(location) {
  companion object {
    @JvmStatic
    val T = Bool(true)
    @JvmStatic
    val Nil = Bool(false)
  }

  constructor(boolean: Boolean) : this(0, boolean)

  override fun recognize() = NodeType.BOOL

  override fun toString() = if (bool) "T" else "NIL"

  override fun equals(other: Any?) = other is Bool && other.bool == bool

  override fun hashCode() = bool.hashCode()

  @Throws(FatalException::class)
  override fun and(location: Int, rhs: Node): Bool {
    if (rhs is Bool) return if (bool && rhs.bool) Bool.T else Bool.Nil
    throw FatalException(location, 117)
  }

  @Throws(FatalException::class)
  override fun or(location: Int, rhs: Node): Bool {
    if (rhs is Bool) return if (bool || rhs.bool) Bool.T else Bool.Nil
    throw FatalException(location, 118)
  }

  @Throws(FatalException::class)
  override fun xor(location: Int, rhs: Node): Bool {
    if (rhs is Bool) return if (bool xor rhs.bool) Bool.T else Bool.Nil
    throw FatalException(location, 119)
  }
}
